File: RoomExample/app/src/main/java/ppjh/mvvm/roomexample/MainViewModel.java
package ppjh.mvvm.roomexample;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.room.Room;

import java.util.List;

/**
 * Context가 필요한 ViewModel은 AndroidViewModel을 상속받아 사
 */
public class MainViewModel extends AndroidViewModel {
    AppDatabase db;
    public String newTodo;
    public LiveData<List<Todo>> todos;

    public MainViewModel(@NonNull Application application) {
        super(application);
        db = Room.databaseBuilder(application, AppDatabase.class, "todo-db")
                .allowMainThreadQueries()
                .build();
        todos = getAll();
    }

    public LiveData<List<Todo>> getAll() {
        return db.todoDao().getAll();
    }
    
    public void insert(String todo) {
        db.todoDao().insert(new Todo(todo));
    }
}

File: RoomExampleKotlin/app/src/main/java/ppjh/mvvm/roomexamplekotlin/MainViewModel.kt
package ppjh.mvvm.roomexamplekotlin

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import androidx.room.Room
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class MainViewModel(application: Application): AndroidViewModel(application) {
    private val db = Room.databaseBuilder(application, AppDatabase::class.java, "todo-db")
        .allowMainThreadQueries()
        .build()

    var todos: LiveData<List<Todo>>
    var newTodo: String? = null

    init {
        todos = getAll()
    }

    fun getAll(): LiveData<List<Todo>> {
        return db.todoDao().getAll()
    }

    //suspend keyword로 coroutine 스코프 안에서 처리되도록 강제
    fun insert(todo: String) {
        viewModelScope.launch(Dispatchers.IO) { db.todoDao().insert(Todo(todo)) }
    }

}